src/build.d

```diff
@@ -475,6 +482,137 @@ alias dmdDefault = makeRule!((builder, rule) => builder
     .description("Build dmd")
     .deps([dmdExe(null, null, null), dmdConf])
 );
+struct PGOState
+{
+    //Does the host compiler actually support PGO, if not print a message
+    static bool checkPGO(string x)
+    {
+        switch (env["HOST_DMD_KIND"])
+        {
+            case "dmd":
+                abortBuild(`DMD does not support PGO!`);
+                break;
+            case "ldc":
+                return true;
+                break;
+            case "gdc":
+                abortBuild(`PGO (or AutoFDO) builds are not yet supported for gdc`);
+                break;
+            default:
+                assert(false, "Unknown host compiler kind: " ~ env["HOST_DMD_KIND"]);
+        }
+        assert(0);
+    }
+    this(string set)
+    {
+        hostKind = set;
+    }
+    string hostKind;
+    string[] pgoGenerateFlags() const
+    {
+        switch(hostKind)
+        {
+            case "ldc":
+                return ["-fprofile-instr-generate=" ~ pgoDataPath ~ "/data.%p.raw"];
+            default:
+                return [""];
+        }
+    }
+    string[] pgoUseFlags() const
+    {
+        switch(hostKind)
+        {
+            case "ldc":
+                return ["-fprofile-instr-use=merged.data"];
+            default:
+                return [""];
+        }
+    }
+    string pgoDataPath() const
+    {
+        const path = buildPath(env["G"], "dmd_profdata");
+        mkdirRecurse(path);
+        return path;
+    }
+}
+//__gshared PGOState pgoState;
+alias dmdPGO = makeRule!((builder, rule) {
+    const dmdKind = env["HOST_DMD_KIND"];
+    PGOState pgoState = PGOState(dmdKind);
+
+    alias buildInstrumentedDmd = methodInit!(BuildRule, (rundBuilder, rundRule) => rundBuilder
+        .msg("Build dmd with profile-guided optimization enabled")
+        .deps([dmdExe(null, pgoState.pgoGenerateFlags(), pgoState.pgoGenerateFlags()), dmdConf]));
+    alias testRunner = methodInit!(BuildRule, (rundBuilder, rundRule) => rundBuilder
+        .msg("(DC) RUN.D")
+        .sources([ testDir.buildPath( "run.d") ])
+        .target(env["GENERATED"].buildPath("run".exeName))
+        .command([ env["HOST_DMD_RUN"], "-of=" ~ rundRule.target, "-i", "-I" ~ testDir] ~ rundRule.sources));
+
+    alias genData = methodInit!(BuildRule, (rundBuilder, rundRule) => rundBuilder
+        .msg("PGO data generated")
+        .sources([ testDir.buildPath( "run.d") ])
+        .deps([buildInstrumentedDmd, testRunner])
+        .commandFunction({
+            // Run dmd test suite to get data
+            const scope cmd = [ testRunner.targets[0], "compilable", "-j" ~ jobs.to!string ];
```

Comment:
Why restrict this to `compilable` tests?

Reply:
So it runs in less than 20 years. Why would you run the tests?

Reply:
More balanced coverage data, especially for the backend (quite a few compilable tests use `-o-`). Running `runnable compilable` (and maybe `fail_compilation`) is fast when omitting the useless permutations for `-O -inline ...` (disabled by passing `ARGS=` to `run.d`)